package/MDAnalysis/analysis/hydrogenbonds/hbond_analysis.py

```diff
@@ -679,90 +679,136 @@ def _get_dh_pairs(self):
+                        f"{self.d_a_cutoff} between Donor, {self.donors_sel}, and "
+                        f"Acceptor, {self.acceptors_sel}."
+                    )
+
+                
+                tmp_donors = donors_for_dist[d_a_indices_12.T[0]]
+                tmp_acceptors = acceptors_for_dist[d_a_indices_12.T[1]]
+                distance = d_a_distances_12
+
+                # Fetch hydrogens bonded to the tmp_donorss
+                # TODO: figure out vectorization
+                for i in range(len(d_a_indices_12)):
+                    if tmp_donors[i].index in donor_idx_to_H_indices:
+                        possible_H_indices = donor_idx_to_H_indices[tmp_donors[i].index]
+                        for h_idx in possible_H_indices:
+                            tmp_hbond_candidates.append(
+                                (tmp_donors[i].index, h_idx, tmp_acceptors[i].index, distance[i])
+                            )
+
+                if np.array_equal(group1.indices, group2.indices):
+                    continue # in case between 2 same residues like ["protein", "protein"]
+
+                tmp_donors_g2_mask = np.isin(self._donors.indices, group2.indices) 
+                donors_for_dist = self._donors[tmp_donors_g2_mask] 
+                tmp_acceptors_g1_mask = np.isin(self._acceptors.indices, group1.indices) 
+                acceptors_for_dist = self._acceptors[tmp_acceptors_g1_mask] 
+
+                d_a_indices_21, d_a_distances_21 = capped_distance(
+                    donors_for_dist.positions,
+                    acceptors_for_dist.positions,
+                    max_cutoff=self.d_a_cutoff,
+                    box=box,
+                    return_distances=True,
+                )
+
+                if np.size(d_a_indices_12) == 0:
+                    warnings.warn(
+                        "No hydrogen bonds were found given d-a cutoff of "
+                        f"{self.d_a_cutoff} between Donor, {self.donors_sel}, and "
+                        f"Acceptor, {self.acceptors_sel}."
+                    )
+                
+
+                tmp_donors = donors_for_dist[d_a_indices_21.T[0]]
+                tmp_acceptors = acceptors_for_dist[d_a_indices_21.T[1]]
+                distance = d_a_distances_21
+
+                for i in range(len(d_a_indices_21)):
+                    if tmp_donors[i].index in donor_idx_to_H_indices:
+                        possible_H_indices = donor_idx_to_H_indices[tmp_donors[i].index]
+                        for h_idx in possible_H_indices:
+                            tmp_hbond_candidates.append(
+                                (tmp_donors[i].index, h_idx, tmp_acceptors[i].index, distance[i])
+                            )
+
+            unique_hbond_dict = {}
+            for d_idx, h_idx, a_idx, dist in tmp_hbond_candidates:
+                key = (d_idx, h_idx, a_idx)
+                if key not in unique_hbond_dict:
+                    unique_hbond_dict[key] = dist
```

Comment:
I think if you do `tmp_hydrogens = self._hydrogens[d_a_indices.T[0]]` as suggested above, then all the hydrogen bonds will be unique anyway and you won't need this dictionary